nvflare/app_common/executors/multi_process_executor.py

```diff
@@ -312,14 +309,14 @@ def finalize(self, fl_ctx: FLContext):
         self.finalized = True
         self.stop_execute = True
 
-        data = {CommunicationMetaData.COMMAND: CommunicateData.CLOSE}
-        for conn_client in self.conn_clients:
-            try:
-                conn_client[CommunicationMetaData.EXE_CONN].send(data)
-                conn_client[CommunicationMetaData.HANDLE_CONN].send(data)
-                self.logger.info("close command sent to rank processes.")
-            except:
-                self.logger.warning("Failed to send the close command. ")
+        request = new_cell_message({}, None)
+        self.engine.client.cell.fire_and_forget(
+            targets=self.targets,
+            channel=CellChannel.CLIENT_SUB_WORKER_COMMAND,
+            topic=MultiProcessCommandNames.CLOSE,
+            message=request,
+        )
+
```

Comment:
The old code is blocking logic VS the new code is non-blocking.

This will cause the next line `os.killpg(os.getpgid(self.exe_process.pid), 9)` be executed right away?

Reply:
This is only to notify the sub_worker_process to finish running. No need to have it in the blocking way.

Reply:
Got it, currently this is fine.
In the future, if the sub_worker_process have some kind of event firing that are queued and we have the mechanism of sending them back, we might need to wait for those events to finished before we go to next line and kill all worker processes.